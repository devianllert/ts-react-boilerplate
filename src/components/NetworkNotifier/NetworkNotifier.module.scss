@import '../../design/design-system';

$notifier-height: 30px;

.notifier {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;

  display: flex;

  align-items: center;
  justify-content: center;

  width: 100%;
  height: $notifier-height;

  transition: all 0.5s ease
}

.notifierMessage {
  color: var(--brand-text-color-inverse);

  font-size: $font-size;
}

.notifierError {
  background-color: var(--brand-danger);
}

.notifierSuccess {
  background-color: var(--brand-success);
}

.notifierClose {
  position: absolute;
  top: 0;
  right: 0;

  display: flex;

  align-items: center;

  height: 100%;
}

.slideAppear {
  transform: translateY($notifier-height);
}

.slideEnter {
  transform: translateY($notifier-height);
}

.slideEnterActive {
  transform: translateY(0);

  transition: transform 250ms;
}

.slideExit {
  transform: translateY(0);
}

.slideExitActive {
  transform: translateY($notifier-height);

  transition: transform 250ms;
}