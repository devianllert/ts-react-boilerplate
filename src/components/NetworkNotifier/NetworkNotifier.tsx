import React, {
  useState,
  useRef,
  ReactElement,
  useEffect,
} from 'react';
import { useTranslation } from 'react-i18next';
import { CSSTransition } from 'react-transition-group';
import classnames from 'classnames';

import useNetwork from '../../hooks/useNetwork';

import Button from '../Button';

import styles from './NetworkNotifier.module.scss';

const NetworkNotifier = (): ReactElement | null => {
  const [show, setShow] = useState(false);
  const timer = useRef<number>();
  const { t } = useTranslation();
  const network = useNetwork((state): void => {
    setShow(true);

    if (state.online) {
      timer.current = window.setTimeout((): void => setShow(false), 5000);
    }
  });

  useEffect((): (() => void) => (): void => {
    if (timer.current) {
      clearTimeout(timer.current);
    }
  }, []);

  const closeNotifier = (): void => {
    if (timer.current) {
      clearTimeout(timer.current);
    }

    setShow(false);
  };

  const message = network.online
    ? t('common.networkNotifier.connected')
    : t('common.networkNotifier.disconnected');

  const classes = classnames(
    styles.notifier,
    network.online ? styles.notifierSuccess : styles.notifierError,
  );

  return (
    <CSSTransition
      timeout={250}
      in={show}
      mountOnEnter
      unmountOnExit
      classNames={{
        enter: styles.slideEnter,
        enterActive: styles.slideEnterActive,
        exit: styles.slideExit,
        exitActive: styles.slideExitActive,
      }}
    >
      <div className={classes}>
        <div className={styles.notifierClose}>
          <Button
            size="small"
            appearence="default"
            flat
            onClick={closeNotifier}
          >
            {t('common.networkNotifier.close')}
          </Button>
        </div>

        <span className={styles.notifierMessage}>{message}</span>
      </div>
    </CSSTransition>
  );
};

export default NetworkNotifier;
