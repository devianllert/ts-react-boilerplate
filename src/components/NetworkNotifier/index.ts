import NetworkNotifier from './NetworkNotifier';

export default NetworkNotifier;
