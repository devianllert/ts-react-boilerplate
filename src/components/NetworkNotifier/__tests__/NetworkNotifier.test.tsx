import React, { ReactElement } from 'react';
import { cleanup, render } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';

import NetworkNotifier from '../NetworkNotifier';

jest.mock('react-transition-group', (): any => {
  const FakeTransition = jest.fn(({ children }): any => children);

  const FakeCSSTransition = jest.fn(
    (props: any): ReactElement | null => <FakeTransition>{props.children}</FakeTransition>,
  );

  return { CSSTransition: FakeCSSTransition, Transition: FakeTransition };
});

describe('<NetworkNotifier />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const LinkButton = render(<NetworkNotifier />);

    expect(LinkButton.container.firstChild).toMatchSnapshot();
  });
});
