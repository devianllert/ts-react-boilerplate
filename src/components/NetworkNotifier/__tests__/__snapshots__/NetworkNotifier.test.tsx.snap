// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NetworkNotifier /> should render and match the snapshot 1`] = `
<div
  class="notifier notifierError"
>
  <div
    class="notifierClose"
  >
    <button
      class="button default small flat"
      type="button"
    >
      common.networkNotifier.close
    </button>
  </div>
  <span
    class="notifierMessage"
  >
    common.networkNotifier.disconnected
  </span>
</div>
`;
